[processor/transform] Add Conversion Function to OTTL for Exponential Histo --> Histogram

daidokoro opened the pull request:
## Description

This PR adds a custom metric function to the transformprocessor to convert exponential histograms to explicit histograms.

Link to tracking issue: Resolves #33827

**Function Name**
```
convert_exponential_histogram_to_explicit_histogram
```

**Arguments:**

- `distribution` (_upper, midpoint, uniform, random_)
- `ExplicitBoundaries: []float64`

**Usage example:**

```yaml
processors:
  transform:
    error_mode: propagate
    metric_statements:
    - context: metric
      statements:
        - convert_exponential_histogram_to_explicit_histogram("random", [10.0, 20.0, 30.0, 40.0, 50.0, 60.0, 70.0, 80.0, 90.0, 100.0]) 
```

**Converts:**

```
Resource SchemaURL: 
ScopeMetrics #0
ScopeMetrics SchemaURL: 
InstrumentationScope  
Metric #0
Descriptor:
     -> Name: response_time
     -> Description: 
     -> Unit: 
     -> DataType: ExponentialHistogram
     -> AggregationTemporality: Delta
ExponentialHistogramDataPoints #0
Data point attributes:
     -> metric_type: Str(timing)
StartTimestamp: 1970-01-01 00:00:00 +0000 UTC
Timestamp: 2024-07-31 09:35:25.212037 +0000 UTC
Count: 44
Sum: 999.000000
Min: 40.000000
Max: 245.000000
Bucket (32.000000, 64.000000], Count: 10
Bucket (64.000000, 128.000000], Count: 22
Bucket (128.000000, 256.000000], Count: 12
        {"kind": "exporter", "data_type": "metrics", "name": "debug"}
```

**To:**

```
Resource SchemaURL: 
ScopeMetrics #0
ScopeMetrics SchemaURL: 
InstrumentationScope  
Metric #0
Descriptor:
     -> Name: response_time
     -> Description: 
     -> Unit: 
     -> DataType: Histogram
     -> AggregationTemporality: Delta
HistogramDataPoints #0
Data point attributes:
     -> metric_type: Str(timing)
StartTimestamp: 1970-01-01 00:00:00 +0000 UTC
Timestamp: 2024-07-30 21:37:07.830902 +0000 UTC
Count: 44
Sum: 999.000000
Min: 40.000000
Max: 245.000000
ExplicitBounds #0: 10.000000
ExplicitBounds #1: 20.000000
ExplicitBounds #2: 30.000000
ExplicitBounds #3: 40.000000
ExplicitBounds #4: 50.000000
ExplicitBounds #5: 60.000000
ExplicitBounds #6: 70.000000
ExplicitBounds #7: 80.000000
ExplicitBounds #8: 90.000000
ExplicitBounds #9: 100.000000
Buckets #0, Count: 0
Buckets #1, Count: 0
Buckets #2, Count: 0
Buckets #3, Count: 2
Buckets #4, Count: 5
Buckets #5, Count: 0
Buckets #6, Count: 3
Buckets #7, Count: 7
Buckets #8, Count: 2
Buckets #9, Count: 4
Buckets #10, Count: 21
        {"kind": "exporter", "data_type": "metrics", "name": "debug"}
```

### Testing

- Several unit tests have been created. We have also tested by ingesting and converting exponential histograms from the `statsdreceiver` as well as directly via the `otlpreceiver` over grpc over several hours with a large amount of data.

- We have clients that have been running this solution in production for a number of weeks.

### Readme description:

### convert_exponential_hist_to_explicit_hist

`convert_exponential_hist_to_explicit_hist([ExplicitBounds])`

the `convert_exponential_hist_to_explicit_hist` function converts an ExponentialHistogram to an Explicit (_normal_) Histogram.

`ExplicitBounds` is represents the list of bucket boundaries for the new histogram. This argument is __required__ and __cannot be empty__.

__WARNING:__

The process of converting an ExponentialHistogram to an Explicit Histogram is not perfect and may result in a loss of precision. It is important to define an appropriate set of bucket boundaries to minimize this loss. For example, selecting Boundaries that are too high or too low may result histogram buckets that are too wide or too narrow, respectively.

daidokoro committed:
added convert_exponential_hist_to_bucketed_hist function

daidokoro committed:
added feature flag for new convert function

daidokoro committed:
added tests for convert_exponential_hist_to_bucketed_hist

daidokoro committed:
removed commented code

daidokoro committed:
add changelog

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
renamed function

adjusted exponential conversion algorithm

daidokoro committed:
renamed and fixed test

daidokoro committed:
changed function name

daidokoro committed:
updated README

daidokoro committed:
readme

daidokoro committed:
added warning about usage

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
add more testx

daidokoro committed:
updated function comments

daidokoro committed:
Merge branch 'cds-1320' of github.com:coralogix/opentelemetry-collector-contrib into cds-1320

daidokoro committed:
updated feature-gate flag name

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
adjust feature gate description

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
fixed typos

daidokoro committed:
Merge branch 'cds-1320' of github.com:coralogix/opentelemetry-collector-contrib into cds-1320

daidokoro committed:
added chloggen

evan-bradley commented:
Hi, @daidokoro, thanks for working on this. I don't have the required expertise with histograms to give this a meaningful review from that perspective, but I've requested for someone who does to take a look at this. I think there may need to be additional discussion on the issue before we can start looking at the code in-depth.

daidokoro committed:
removed duplicate change log

dyladan commented:
Hi there. This looks reasonable as a approximation to me, but I would be worried about introducing bias in the translation. This looks to be taking the upper bound of each bucket and using that as the value for the points in that bucket. So if a bucket is from `1` to `2`, this is just assuming a value of `2` and inserting it into the destination histogram. I think this has the potential to introduce significant bias, especially on larger values which are in wider exponential buckets.

To me, there are several other possible strategies that would work better:

- choose the midpoint of the bucket. This may also introduce some bias in some cases, but I think in the general case it is likely to be better than choosing the upper bound of the bucket
- uniformly spread the points in the bucket throughout the bucket range
- randomly spread the points in the bucket throughout the bucket range, either uniformly or with some other distribution

There are other strategies but these would be good to start with. Personally, I would prefer a way for the user to choose between the strategies and have the default be something sensible that balances performance and complexity like uniform or uniform random distribution.

daidokoro commented:
> Hi there. This looks reasonable as a approximation to me, but I would be worried about introducing bias in the translation. This looks to be taking the upper bound of each bucket and using that as the value for the points in that bucket. So if a bucket is from `1` to `2`, this is just assuming a value of `2` and inserting it into the destination histogram. I think this has the potential to introduce significant bias, especially on larger values which are in wider exponential buckets.
> 
> To me, there are several other possible strategies that would work better:
> 
> * choose the midpoint of the bucket. This may also introduce some bias in some cases, but I think in the general case it is likely to be better than choosing the upper bound of the bucket
> * uniformly spread the points in the bucket throughout the bucket range
> * randomly spread the points in the bucket throughout the bucket range, either uniformly or with some other distribution
> 
> There are other strategies but these would be good to start with. Personally, I would prefer a way for the user to choose between the strategies and have the default be something sensible that balances performance and complexity like uniform or uniform random distribution.

Thanks for the feedback @dyladan 

I agree that the conversion using the upper bound can lead to a particular bias in some cases. Having alternative approaches could help users select the best conversion approach based on their metric data.

I will look at adding requested functionality, i.e
- Conversion function using the mean/average of know upper and lower bound
- Conversion function using uniformly spread points
- Conversion function using random spread points
- An option for users to select the approach they want to use to convert.

Thanks.

daidokoro committed:
updated readme with addition distribution approaches

daidokoro committed:
added random, uniform and midpoint distribution functions and updated functionf flow to support

daidokoro committed:
added tests for random, uniform and midpoint distribution implementations

daidokoro committed:
added overflow assertion to random dist test

daidokoro committed:
updated readme

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro commented:
Hey @dyladan , 

I've added the strategies you requested as well as a way for users to select which approach they want to use.
- upper
- midpoint
- uniform
- random

Rather than add an explicit default, the user must select the strategy they want to use, that way users will atleast engage with the various strategies and find what fits their data best.

When possible, let me know if anything is required here.

Thanks.

kentquirk commented on processor/transformprocessor/README.md:
```diff
@@ -355,6 +357,39 @@ Examples:
+<br>
+  - __upper__ - This approach identifies the highest possible value of each exponential bucket (_the upper bound_) and uses it to distribute the datapoints by comparing the upper bound of each bucket with the ExplicitBounds provided. This approach works better for small/narrow exponential histograms where the difference between the upper bounds and lower bounds are small.
+<br>
+  - __midpoint__ - This approach works in a similar way to the __upper__ approach, but instead of using the upper bound, it uses the midpoint of each exponential bucket. The midpoint is identified by calculationg the average of the upper and lower bounds. This approach also works better for small/narrow exponential histograms.
```
```suggestion
  - __midpoint__ - This approach works in a similar way to the __upper__ approach, but instead of using the upper bound, it uses the midpoint of each exponential bucket. The midpoint is identified by calculating the average of the upper and lower bounds. This approach also works better for small/narrow exponential histograms.
```

kentquirk commented on processor/transformprocessor/README.md:
```diff
@@ -355,6 +357,39 @@ Examples:
+<br>
+  - __midpoint__ - This approach works in a similar way to the __upper__ approach, but instead of using the upper bound, it uses the midpoint of each exponential bucket. The midpoint is identified by calculationg the average of the upper and lower bounds. This approach also works better for small/narrow exponential histograms.
+<br>
+  - __uniform__ - This approach distributes the datapoints for each bucket uniformly across the overlapping __ExplicitBounds__. This approach works better for large/wide exponential histograms.
```
"overlapping" doesn't say enough about what you mean here. I can't tell whether you're generating values uniformly between the exponential histogram's bucket bounds, or between the explicit bounds.

The same is true for the next bullet.

daidokoro replied:
Thanks, added a more detailed description of what these distribution algorithms are doing.

kentquirk commented on processor/transformprocessor/README.md:
```diff
@@ -355,6 +357,39 @@ Examples:
+  - __uniform__ - This approach distributes the datapoints for each bucket uniformly across the overlapping __ExplicitBounds__. This approach works better for large/wide exponential histograms.
+<br>
+  - __random__ - This approach distributes the datapoints for each bucket randomly across the overlapping __ExplicitBounds__. This approach works better for large/wide exponential histograms.
+<br>
```
I think an example showing maybe 2 input buckets and the set of values that each of these distributions would put into the output would go a very long way.

daidokoro replied:
Thanks, Added examples outlining the distribution process

TylerHelmuth commented on processor/transformprocessor/internal/metrics/functions.go:
```diff
@@ -18,6 +18,12 @@ var useConvertBetweenSumAndGaugeMetricContext = featuregate.GlobalRegistry().Mus
 	featuregate.WithRegisterDescription("When enabled will use metric context for conversion between sum and gauge"),
 )
 
+var useConvertExponentialHistogramToExplicitHistogram = featuregate.GlobalRegistry().MustRegister(
```
Since a user must explicitly configure the function if they want the conversion, we do not need to block the feature via a feature gate.

daidokoro replied:
Thanks, Feature gate removed.

kentquirk reviewed (changes requested):
I left some specific comments, but we need a bit of a reorganization here.

First, the extraction of points from the exponential histogram can be centralized -- there's no need to repeat the code that iterates over the incoming buckets and finds the bounds for each one. The difference should be in how you distribute, say, 10 values in the range from 2 to 4 among the outgoing buckets. 

Note that if an incoming bucket falls entirely within a single outgoing bucket (which it often will), the distribution algorithm doesn't matter at all and we could take a shortcut path to just incrementing the count in the outgoing bucket.

Once you have more than one outgoing bucket in hand, you can choose the algorithm for how the counts are distributed among them and do the math then.

Second, this code doesn't account for the ZeroCount part of the exponential histogram. (It also doesn't do Negatives, but last I looked, nothing else does them either.)

Finally, it doesn't copy over any Exemplars from the source to the destination. I believe it should.

daidokoro committed:
fix default

daidokoro committed:
update default

daidokoro committed:
removed featuregate

daidokoro committed:
add function to test

daidokoro committed:
updated tests

daidokoro committed:
- updated algorithms, fixed minor issues with accuracy of conversion

daidokoro committed:
updated readme

daidokoro committed:
Update processor/transformprocessor/README.md

Co-authored-by: Kent Quirk <kentquirk@gmail.com>

daidokoro commented:
Thanks @kentquirk ,

I've addressed your recommendations below.

Let me know if anything else is needed.

> 
> First, the extraction of points from the exponential histogram can be centralized -- there's no need to repeat the code that iterates over the incoming buckets and finds the bounds for each one. The difference should be in how you distribute, say, 10 values in the range from 2 to 4 among the outgoing buckets.

- Centralised the distribution functions and reduced unnecessary repetition.

> Note that if an incoming bucket falls entirely within a single outgoing bucket (which it often will), the distribution algorithm doesn't matter at all and we could take a shortcut path to just incrementing the count in the outgoing bucket.
> 
> Once you have more than one outgoing bucket in hand, you can choose the algorithm for how the counts are distributed among them and do the math then.

- Modified flow to check if lower and upper bounds fit entirely within one bucket and allocate counts directly 
- Modified flow to check if $lower Bound > largestExplicitBoundary$, and allocate the count directly to the overflow bucket
- Only if the above 2 conditions are _not_ met will the distribution algorithms be applied.
 

> Second, this code doesn't account for the ZeroCount part of the exponential histogram. (It also doesn't do Negatives, but last I looked, nothing else does them either.)
- Added support for ZeroCount when 0 boundary is present in user provided Explicit Boundary
- Added description of this behaviour to the README.md
- Added a note to the README.md regarding Negative Buckets Counts not being supported.

> Finally, it doesn't copy over any Exemplars from the source to the destination. I believe it should.
- Updated to copy Exemplars

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

kentquirk reviewed (approved):
This is impressive work. I'm now very happy with this.

TylerHelmuth commented on processor/transformprocessor/README.md:
```diff
@@ -355,6 +357,82 @@ Examples:
+
+`convert_exponential_hist_to_explicit_hist(distribution, [ExplicitBounds])`
+
+The `convert_exponential_hist_to_explicit_hist` function converts an ExponentialHistogram to an Explicit (_normal_) Histogram.
```
Please add before this line another warning stating this behavior is not defined by the opentelemetry spec.

daidokoro replied:
Added the following:

__Warning:__ The approach used in this function to convert exponential histograms to explicit histograms __is not__ part of the __OpenTelemetry Specification__.

TylerHelmuth commented on processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist.go:
```diff
@@ -0,0 +1,305 @@
+}
+
+func newconvertExponentialHistToExplicitHistFactory() ottl.Factory[ottlmetric.TransformContext] {
+	return ottl.NewFactory("convert_exponential_hist_to_explicit_hist",
```
lets use exact otel terms in the name: `convert_exponential_histogram_to_histogram`

daidokoro replied:
Function name updated to: `convert_exponential_histogram_to_explicit_histogram`

TylerHelmuth replied:
The otel spec doesn't use the term `explicit` in the name as far as I know.  Using only `histogram` should be sufficient.

daidokoro replied:
Removed

TylerHelmuth commented on processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist.go:
```diff
@@ -0,0 +1,305 @@
+
+	distFn, ok := distributionFnMap[distributionFn]
+	if !ok {
+		return nil, fmt.Errorf("invalid distribution alogrithm: %s, must be one of [upper, midpoint, random, uniform]", distributionFn)
```
```suggestion
		return nil, fmt.Errorf("invalid distribution algorithm: %s, must be one of [upper, midpoint, random, uniform]", distributionFn)
```

TylerHelmuth commented on processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist.go:
```diff
@@ -0,0 +1,305 @@
+		bucketCounts[0] += zerocount
+	}
+
+positionLoop:
+	for pos := 0; pos < posB.Len(); pos++ {
+		index := dp.Positive().Offset() + int32(pos)
+		upper := math.Exp(float64(index+1) * factor)
+		lower := math.Exp(float64(index) * factor)
+		count := posB.At(pos)
+
+		// check if lower and upper bounds are within the boundaries
+		for bIndex := 1; bIndex < len(boundaries); bIndex++ {
+			if lower > boundaries[bIndex-1] && upper <= boundaries[bIndex] {
+				bucketCounts[bIndex-1] += count
+				continue positionLoop
+			}
+		}
```
We have a design principle (technically for standard OTTL functions, which this is not, but I still like to point to them) that requires that OTTL functions be [terminating](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md#design-principles).  Is there a way to refactor this loop to not need a GOTO? The value instead the if statement isn't used in the condition, so is it possible to get stuck continuously being sent back to the start of the first for loop?

daidokoro replied:
Wrapped the `distFn` call in a condition, changed the `continue` to `break` and removed the GOTO

daidokoro committed:
Update processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist.go

fix typo

Co-authored-by: Tyler Helmuth <12352919+TylerHelmuth@users.noreply.github.com>

daidokoro committed:
update function name and add warning message

daidokoro committed:
fixed function name

daidokoro committed:
remove GOTO from distFn loop and changed transform function name

daidokoro commented:
Thanks @TylerHelmuth, the requested changes have been committed.

daidokoro committed:
adjust function name

TylerHelmuth reviewed (approved):
The OTTL bits look correct. I cannot verify the math.

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
fix typos

daidokoro committed:
Merge branch 'cds-1320' of github.com:coralogix/opentelemetry-collector-contrib into cds-1320

daidokoro committed:
fix linting issues

daidokoro committed:
Merge branch 'main' into cds-1320

evan-bradley commented:
@daidokoro Really appreciate your continued efforts on this. I am also not able to assess whether these changes will produce correct results, but I have requested for someone who can to take a look. I'll give a review from the OTTL side after that and hopefully we can get this merged.

dyladan reviewed (approved):
Sorry it took so long for me to review. This looks good and I appreciate you adding more options for the conversions.

daidokoro committed:
Merge branch 'main' into cds-1320

TylerHelmuth commented:
@daidokoro looks like you need a `go mod tidy` and then we'll be good.

daidokoro committed:
go mod tidy on transformprocessor

daidokoro commented:
> @daidokoro looks like you need a `go mod tidy` and then we'll be good.

Thanks @TylerHelmuth , executed.

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
go mod tidy

daidokoro committed:
fixed linting issue

TylerHelmuth commented on processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist_test.go:
```diff
@@ -0,0 +1,787 @@
+	"testing"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
+	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
```
```suggestion
	"github.com/stretchr/testify/assert"
	"go.opentelemetry.io/collector/pdata/pcommon"
	"go.opentelemetry.io/collector/pdata/pmetric"
	
	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
```

daidokoro committed:
Update processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist_test.go

Co-authored-by: Tyler Helmuth <12352919+TylerHelmuth@users.noreply.github.com>

TylerHelmuth commented on processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist_test.go:
```diff
@@ -0,0 +1,788 @@
+	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	
```
```suggestion

```

daidokoro committed:
Update processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist_test.go

Co-authored-by: Tyler Helmuth <12352919+TylerHelmuth@users.noreply.github.com>

daidokoro committed:
go fmt

daidokoro committed:
Merge branch 'main' into cds-1320

daidokoro committed:
gci-ed

TylerHelmuth commented:
Another thing to do is to run `make checks` locally. Make sure your local Go version is `1.22.7`.

daidokoro commented:
> @daidokoro this has been a bad week for our CI as there has been a lot of churning in Core that is causing us to do more core library updates than normal. In addition some new linters are causing issues and the prometheus compliance tests as well. Also the general Go proxy issue.
> 
> Things are getting cleaned up. One thing that would help is if you allow us to merge commits into your branch from this PR. That would allow me to update your branch with the latest from main and merge code suggestions I make.

Thanks @TylerHelmuth for the update. Feel free to make commits and I'll merge all suggestions. Anything else I can do to help let me know.

TylerHelmuth commented:
@daidokoro I don't have permission to merge in.  I'd say get the latest from main (it should have all the CI fixed), and then ensure `make checks`

daidokoro committed:
Merge branch 'main' into cds-1320

TylerHelmuth commented:
Needs a `make crosslink`

daidokoro committed:
make crosslink
